File: src/cpp/qrng/ibm_quantum.cpp
/**
 * IBM Quantum QRNG Implementation
 */

#include "ibm_quantum.h"
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>
#include <cstring>
#include <iostream>
#include <sstream>

namespace qrng {

IBMQuantumQRNG::IBMQuantumQRNG(const IBMQuantumConfig& config)
    : config_(config),
      pool_bytes_read_(0),
      pool_total_size_(0),
      initialized_(false),
      using_quantum_(false),
      pool_exhausted_(false),
      urandom_fd_(-1) {

    last_health_check_ = std::chrono::steady_clock::now();
}

IBMQuantumQRNG::~IBMQuantumQRNG() {
    disconnect();
}

QRNGStatus IBMQuantumQRNG::initialize() {
    if (initialized_) {
        return QRNGStatus::OK;
    }

    // Open entropy pool file
    QRNGStatus status = open_pool_file();
    if (status != QRNGStatus::OK) {
        std::cerr << "Warning: Failed to open IBM Quantum entropy pool, using fallback\n";
        using_quantum_ = false;
    } else {
        using_quantum_ = true;
        log_source("IBM Quantum Pool");
    }

    // Open /dev/urandom for fallback
    urandom_fd_ = open("/dev/urandom", O_RDONLY);
    if (urandom_fd_ < 0) {
        std::cerr << "Error: Failed to open /dev/urandom\n";
        return QRNGStatus::DEVICE_ERROR;
    }

    initialized_ = true;
    return QRNGStatus::OK;
}

bool IBMQuantumQRNG::is_connected() const {
    return initialized_;
}

QRNGStatus IBMQuantumQRNG::get_random_bytes(uint8_t* buffer, size_t length) {
    if (!initialized_) {
        return QRNGStatus::NOT_CONNECTED;
    }

    if (!buffer || length == 0) {
        return QRNGStatus::DEVICE_ERROR;
    }

    QRNGStatus status;

    // Try quantum pool first if available
    if (using_quantum_ && !pool_exhausted_) {
        status = read_from_pool(buffer, length);

        if (status == QRNGStatus::OK) {
            update_stats(length, true);

            // Check if pool is running low
            size_t available = available_bytes();
            if (available < config_.min_pool_bytes && config_.warn_on_fallback) {
                std::cerr << "Warning: Quantum entropy pool low (" << available
                         << " bytes remaining). Consider refilling.\n";
            }

            return QRNGStatus::OK;
        }

        // Pool exhausted, switch to fallback
        if (config_.warn_on_fallback) {
            std::cerr << "Warning: Quantum entropy pool exhausted, falling back to /dev/urandom\n";
        }
        pool_exhausted_ = true;
        using_quantum_ = false;
        log_source("System /dev/urandom (fallback)");
    }

    // Use fallback
    status = read_from_urandom(buffer, length);
    update_stats(length, false);

    return status;
}

QRNGStatus IBMQuantumQRNG::health_check() {
    std::lock_guard<std::mutex> lock(stats_mutex_);

    auto now = std::chrono::steady_clock::now();
    last_health_check_ = now;

    // Check if we have quantum entropy available
    if (!using_quantum_ && !pool_exhausted_) {
        stats_.health_checks_failed++;
        return QRNGStatus::HEALTH_CHECK_FAILED;
    }

    // Check pool file validity if using quantum
    if (using_quantum_ && pool_file_.is_open()) {
        if (!pool_file_.good()) {
            stats_.health_checks_failed++;
            return QRNGStatus::IO_ERROR;
        }
    }

    // Check fallback availability
    if (urandom_fd_ < 0) {
        stats_.health_checks_failed++;
        return QRNGStatus::DEVICE_ERROR;
    }

    stats_.health_checks_passed++;
    return QRNGStatus::OK;
}

HealthStats IBMQuantumQRNG::get_health_stats() const {
    std::lock_guard<std::mutex> lock(stats_mutex_);
    HealthStats current_stats = stats_;

    // Update buffer fill percent
    current_stats.buffer_fill_percent = static_cast<uint32_t>(get_pool_percent());

    // Update last health check timestamp
    auto now = std::chrono::steady_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(
        now - last_health_check_
    ).count();
    current_stats.last_health_check_us = elapsed;

    return current_stats;
}

size_t IBMQuantumQRNG::available_bytes() const {
    std::lock_guard<std::mutex> lock(file_mutex_);

    if (!pool_file_.is_open() || pool_exhausted_) {
        return 0;
    }

    return pool_total_size_ - pool_bytes_read_;
}

std::string IBMQuantumQRNG::get_device_info() const {
    std::stringstream ss;
    ss << "IBM Quantum QRNG";

    if (using_quantum_) {
        ss << " (Quantum Active: " << (pool_total_size_ - pool_bytes_read_)
           << "/" << pool_total_size_ << " bytes)";
    } else {
        ss << " (Fallback: /dev/urandom)";
    }

    return ss.str();
}

void IBMQuantumQRNG::disconnect() {
    if (pool_file_.is_open()) {
        pool_file_.close();
    }

    if (urandom_fd_ >= 0) {
        close(urandom_fd_);
        urandom_fd_ = -1;
    }

    initialized_ = false;
}

bool IBMQuantumQRNG::is_using_quantum() const {
    return using_quantum_ && !pool_exhausted_;
}

double IBMQuantumQRNG::get_pool_percent() const {
    if (pool_total_size_ == 0) {
        return 0.0;
    }

    size_t remaining = pool_total_size_ - pool_bytes_read_;
    return (remaining * 100.0) / pool_total_size_;
}

QRNGStatus IBMQuantumQRNG::reload_pool() {
    std::lock_guard<std::mutex> lock(file_mutex_);

    if (pool_file_.is_open()) {
        pool_file_.close();
    }

    pool_bytes_read_ = 0;
    pool_exhausted_ = false;

    QRNGStatus status = open_pool_file();
    if (status == QRNGStatus::OK) {
        using_quantum_ = true;
        log_source("IBM Quantum Pool (reloaded)");
    }

    return status;
}

// Private methods

QRNGStatus IBMQuantumQRNG::read_from_pool(uint8_t* buffer, size_t length) {
    std::lock_guard<std::mutex> lock(file_mutex_);

    if (!pool_file_.is_open() || pool_file_.eof()) {
        return QRNGStatus::BUFFER_UNDERRUN;
    }

    pool_file_.read(reinterpret_cast<char*>(buffer), length);
    size_t bytes_read = pool_file_.gcount();

    if (bytes_read < length) {
        // Pool exhausted mid-read, fill remainder with fallback
        size_t remaining = length - bytes_read;
        QRNGStatus status = read_from_urandom(buffer + bytes_read, remaining);
        if (status != QRNGStatus::OK) {
            return status;
        }

        pool_exhausted_ = true;
    }

    pool_bytes_read_ += bytes_read;

    return QRNGStatus::OK;
}

QRNGStatus IBMQuantumQRNG::read_from_urandom(uint8_t* buffer, size_t length) {
    if (urandom_fd_ < 0) {
        return QRNGStatus::NOT_CONNECTED;
    }

    ssize_t result = read(urandom_fd_, buffer, length);
    if (result < 0 || static_cast<size_t>(result) != length) {
        return QRNGStatus::IO_ERROR;
    }

    return QRNGStatus::OK;
}

QRNGStatus IBMQuantumQRNG::open_pool_file() {
    // Open pool file
    pool_file_.open(config_.pool_file_path, std::ios::binary | std::ios::in);

    if (!pool_file_.is_open()) {
        std::cerr << "Error: Could not open entropy pool file: "
                  << config_.pool_file_path << "\n";
        return QRNGStatus::NOT_CONNECTED;
    }

    // Get file size
    pool_file_.seekg(0, std::ios::end);
    pool_total_size_ = pool_file_.tellg();
    pool_file_.seekg(0, std::ios::beg);

    if (pool_total_size_ == 0) {
        std::cerr << "Error: Entropy pool file is empty\n";
        pool_file_.close();
        return QRNGStatus::BUFFER_UNDERRUN;
    }

    pool_bytes_read_ = 0;
    pool_exhausted_ = false;

    std::cout << "Loaded IBM Quantum entropy pool: "
              << pool_total_size_ << " bytes available\n";

    return QRNGStatus::OK;
}

void IBMQuantumQRNG::log_source(const std::string& source) {
    if (config_.log_entropy_source) {
        std::cout << "[Entropy Source] " << source << "\n";
    }
}

void IBMQuantumQRNG::update_stats(size_t bytes_read, bool from_quantum) {
    std::lock_guard<std::mutex> lock(stats_mutex_);

    stats_.bytes_generated += bytes_read;

    // Calculate entropy rate (simplified)
    static auto last_time = std::chrono::steady_clock::now();
    auto now = std::chrono::steady_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(
        now - last_time
    ).count();

    if (elapsed > 0) {
        stats_.entropy_rate_bps = (bytes_read * 1000000.0) / elapsed;
    }

    last_time = now;
}

} // namespace qrng

File: crates/zipminator-core/src/qrng/ibm_quantum.rs
//! IBM Quantum QRNG Device
//!
//! Provides quantum entropy from IBM Quantum systems via:
//! 1. Pre-generated entropy pool file (primary, offline)
//! 2. IBM Quantum API (optional, requires authentication)
//! 3. System /dev/urandom (fallback)
//!
//! Features:
//! - File-based entropy pool for offline operation
//! - Automatic fallback when pool exhausted
//! - Health monitoring and logging
//! - Thread-safe concurrent access

use super::{HealthStatus, QrngDevice, QrngError};
use log::{info, warn};
use std::fs::File;
use std::io::Read;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};

/// Configuration for IBM Quantum QRNG
#[derive(Debug, Clone)]
pub struct IBMQuantumConfig {
    /// Path to pre-generated entropy pool file
    pub pool_file_path: PathBuf,
    /// Minimum bytes before warning
    pub min_pool_bytes: usize,
    /// Warn when falling back to /dev/urandom
    pub warn_on_fallback: bool,
    /// Log which entropy source is being used
    pub log_entropy_source: bool,
}

impl Default for IBMQuantumConfig {
    fn default() -> Self {
        Self {
            pool_file_path: PathBuf::from("/var/lib/zipminator/quantum_entropy.pool"),
            min_pool_bytes: 10240, // 10KB
            warn_on_fallback: true,
            log_entropy_source: true,
        }
    }
}

/// Statistics for IBM Quantum QRNG
#[derive(Debug, Default)]
pub struct IBMQuantumStats {
    bytes_served: u64,
    bytes_from_quantum: u64,
    bytes_from_fallback: u64,
    pool_exhausted_count: u64,
}

/// IBM Quantum QRNG Device
pub struct IBMQuantumQRNG {
    config: IBMQuantumConfig,
    pool_file: Option<File>,
    pool_total_size: u64,
    pool_bytes_read: u64,
    urandom_file: Option<File>,
    using_quantum: bool,
    pool_exhausted: bool,
    initialized: bool,
    stats: Arc<Mutex<IBMQuantumStats>>,
}

impl IBMQuantumQRNG {
    /// Create a new IBM Quantum QRNG with default configuration
    pub fn new() -> Self {
        Self::with_config(IBMQuantumConfig::default())
    }

    /// Create a new IBM Quantum QRNG with custom configuration
    pub fn with_config(config: IBMQuantumConfig) -> Self {
        Self {
            config,
            pool_file: None,
            pool_total_size: 0,
            pool_bytes_read: 0,
            urandom_file: None,
            using_quantum: false,
            pool_exhausted: false,
            initialized: false,
            stats: Arc::new(Mutex::new(IBMQuantumStats::default())),
        }
    }

    /// Open and validate the quantum entropy pool file
    fn open_pool_file(&mut self) -> Result<(), QrngError> {
        let path = &self.config.pool_file_path;

        if !path.exists() {
            return Err(QrngError::DeviceNotFound);
        }

        let file = File::open(path).map_err(|e| {
            QrngError::InitializationFailed(format!("Failed to open pool file: {}", e))
        })?;

        // Get file size
        let metadata = file.metadata().map_err(|e| {
            QrngError::InitializationFailed(format!("Failed to read file metadata: {}", e))
        })?;

        self.pool_total_size = metadata.len();

        if self.pool_total_size == 0 {
            return Err(QrngError::InitializationFailed(
                "Entropy pool file is empty".to_string(),
            ));
        }

        info!(
            "Loaded IBM Quantum entropy pool: {} bytes available",
            self.pool_total_size
        );

        self.pool_file = Some(file);
        self.pool_bytes_read = 0;
        self.pool_exhausted = false;

        Ok(())
    }

    /// Open /dev/urandom for fallback
    fn open_urandom(&mut self) -> Result<(), QrngError> {
        let file = File::open("/dev/urandom").map_err(|e| {
            QrngError::InitializationFailed(format!("Failed to open /dev/urandom: {}", e))
        })?;

        self.urandom_file = Some(file);
        Ok(())
    }

    /// Read from quantum entropy pool
    fn read_from_pool(&mut self, buffer: &mut [u8]) -> Result<usize, QrngError> {
        let file = self
            .pool_file
            .as_mut()
            .ok_or(QrngError::DeviceNotFound)?;

        match file.read(buffer) {
            Ok(bytes_read) => {
                if bytes_read == 0 {
                    self.pool_exhausted = true;
                    return Err(QrngError::InsufficientEntropy);
                }

                self.pool_bytes_read += bytes_read as u64;

                // Check if pool is running low
                let remaining = self.pool_total_size - self.pool_bytes_read;
                if remaining < self.config.min_pool_bytes as u64
                    && self.config.warn_on_fallback
                {
                    warn!(
                        "Quantum entropy pool low ({} bytes remaining). Consider refilling.",
                        remaining
                    );
                }

                Ok(bytes_read)
            }
            Err(e) => Err(QrngError::ReadError(format!(
                "Failed to read from pool: {}",
                e
            ))),
        }
    }

    /// Read from /dev/urandom fallback
    fn read_from_urandom(&mut self, buffer: &mut [u8]) -> Result<usize, QrngError> {
        let file = self
            .urandom_file
            .as_mut()
            .ok_or(QrngError::DeviceNotFound)?;

        file.read(buffer)
            .map_err(|e| QrngError::ReadError(format!("Failed to read from urandom: {}", e)))
    }

    /// Log which entropy source is being used
    fn log_source(&self, source: &str) {
        if self.config.log_entropy_source {
            info!("[Entropy Source] {}", source);
        }
    }

    /// Check if currently using quantum entropy
    pub fn is_using_quantum(&self) -> bool {
        self.using_quantum && !self.pool_exhausted
    }

    /// Get percentage of pool remaining
    pub fn get_pool_percent(&self) -> f64 {
        if self.pool_total_size == 0 {
            return 0.0;
        }

        let remaining = self.pool_total_size - self.pool_bytes_read;
        (remaining as f64 / self.pool_total_size as f64) * 100.0
    }

    /// Get statistics
    pub fn get_stats(&self) -> IBMQuantumStats {
        let stats = self.stats.lock().unwrap();
        IBMQuantumStats {
            bytes_served: stats.bytes_served,
            bytes_from_quantum: stats.bytes_from_quantum,
            bytes_from_fallback: stats.bytes_from_fallback,
            pool_exhausted_count: stats.pool_exhausted_count,
        }
    }
}

impl QrngDevice for IBMQuantumQRNG {
    fn initialize(&mut self) -> Result<(), QrngError> {
        if self.initialized {
            return Ok(());
        }

        // Try to open quantum entropy pool
        match self.open_pool_file() {
            Ok(_) => {
                self.using_quantum = true;
                self.log_source("IBM Quantum Pool");
            }
            Err(e) => {
                warn!(
                    "Failed to open IBM Quantum entropy pool: {}. Using fallback.",
                    e
                );
                self.using_quantum = false;
            }
        }

        // Open /dev/urandom for fallback
        self.open_urandom()?;

        self.initialized = true;
        Ok(())
    }

    fn get_random_bytes(&mut self, buffer: &mut [u8]) -> Result<usize, QrngError> {
        if !self.initialized {
            return Err(QrngError::DeviceNotFound);
        }

        let mut total_read = 0;

        // Try quantum pool first if available
        if self.using_quantum && !self.pool_exhausted {
            match self.read_from_pool(buffer) {
                Ok(bytes_read) => {
                    total_read += bytes_read;

                    // Update stats
                    let mut stats = self.stats.lock().unwrap();
                    stats.bytes_served += bytes_read as u64;
                    stats.bytes_from_quantum += bytes_read as u64;

                    // If we got all requested bytes, we're done
                    if bytes_read == buffer.len() {
                        return Ok(total_read);
                    }

                    // Pool exhausted mid-read, fall through to urandom for remainder
                    if self.config.warn_on_fallback {
                        warn!(
                            "Quantum entropy pool exhausted, falling back to /dev/urandom"
                        );
                    }
                    self.pool_exhausted = true;
                    self.using_quantum = false;
                    self.log_source("System /dev/urandom (fallback)");

                    let mut stats = self.stats.lock().unwrap();
                    stats.pool_exhausted_count += 1;
                }
                Err(_) => {
                    // Pool failed, switch to fallback
                    self.pool_exhausted = true;
                    self.using_quantum = false;
                }
            }
        }

        // Use fallback for any remaining bytes
        if total_read < buffer.len() {
            let remaining = &mut buffer[total_read..];
            let bytes_read = self.read_from_urandom(remaining)?;

            // Update stats
            let mut stats = self.stats.lock().unwrap();
            stats.bytes_served += bytes_read as u64;
            stats.bytes_from_fallback += bytes_read as u64;

            total_read += bytes_read;
        }

        Ok(total_read)
    }

    fn health_check(&self) -> Result<HealthStatus, QrngError> {
        if !self.initialized {
            return Err(QrngError::DeviceNotFound);
        }

        // Check if urandom is available (critical)
        if self.urandom_file.is_none() {
            return Ok(HealthStatus::Failed);
        }

        // Check quantum pool status
        if self.using_quantum && !self.pool_exhausted {
            let remaining = self.pool_total_size - self.pool_bytes_read;

            if remaining < self.config.min_pool_bytes as u64 {
                return Ok(HealthStatus::Degraded);
            }

            return Ok(HealthStatus::Healthy);
        }

        // Using fallback, but operational
        Ok(HealthStatus::Degraded)
    }

    fn device_info(&self) -> String {
        if self.is_using_quantum() {
            format!(
                "IBM Quantum QRNG (Quantum Active: {}/{} bytes, {:.1}%)",
                self.pool_total_size - self.pool_bytes_read,
                self.pool_total_size,
                self.get_pool_percent()
            )
        } else {
            "IBM Quantum QRNG (Fallback: /dev/urandom)".to_string()
        }
    }

    fn is_ready(&self) -> bool {
        self.initialized
    }

    fn shutdown(&mut self) -> Result<(), QrngError> {
        self.pool_file = None;
        self.urandom_file = None;
        self.initialized = false;
        Ok(())
    }
}

impl Default for IBMQuantumQRNG {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_config_defaults() {
        let config = IBMQuantumConfig::default();
        assert_eq!(config.min_pool_bytes, 10240);
        assert!(config.warn_on_fallback);
        assert!(config.log_entropy_source);
    }

    #[test]
    fn test_device_creation() {
        let device = IBMQuantumQRNG::new();
        assert!(!device.initialized);
        assert!(!device.using_quantum);
    }

    #[test]
    fn test_pool_percent_calculation() {
        let mut device = IBMQuantumQRNG::new();
        device.pool_total_size = 1000;
        device.pool_bytes_read = 250;

        assert_eq!(device.get_pool_percent(), 75.0);
    }
}
